File: scala/src/test/scala/com/pduda/field/ShoppingCartTest.scala
package com.pduda.field

class ShoppingCartTest extends munit.FunSuite {
  test("single item: number of products in the cart") {
    val shoppingCart = new ShoppingCart
    shoppingCart.add(10)
    assertEquals(1, shoppingCart.numberOfProducts)
  }

  test("single item: total price") {
    val shoppingCart = new ShoppingCart
    shoppingCart.add(10)
    assertEquals(10, shoppingCart.calculateTotalPrice)
  }

  test("single item: has discount if contains at least one product worth at least 100") {
    val shoppingCart = new ShoppingCart
    shoppingCart.add(100)
    assert(shoppingCart.hasDiscount)
  }

  test("single item: does not have discount if contains no products worth at least 100") {
    val shoppingCart = new ShoppingCart
    shoppingCart.add(99)
    assert(!shoppingCart.hasDiscount)
  }
}

File: Java/src/test/java/com/pduda/field/ShoppingCartTest.java
package com.pduda.field;

import org.junit.Assert;
import org.junit.Test;

public class ShoppingCartTest {

    /**
     * Verify that adding a single item results in one product in the cart.
     */
    @Test
    public void singleItemNumberOfProductsInTheCart() throws Exception {
        ShoppingCart shoppingCart = new ShoppingCart();
        shoppingCart.add(10);

        Assert.assertEquals(1, shoppingCart.numberOfProducts());
    }

    /**
     * Verify that the total price reflects the price of the added item.
     */
    @Test
    public void singleItemTotalTotalPrice() throws Exception {
        ShoppingCart shoppingCart = new ShoppingCart();
        shoppingCart.add(10);

        Assert.assertEquals(10, shoppingCart.calculateTotalPrice());
    }

    /**
     * Verify that a discount is applied when at least one product is worth 100
     * or more.
     */
    @Test
    public void singleItemHasDiscountIfContainsAtLeastOneProductWorthAtLeast100() throws Exception {
        ShoppingCart shoppingCart = new ShoppingCart();
        shoppingCart.add(100);

        Assert.assertTrue(shoppingCart.hasDiscount());
    }

    /**
     * Verify that no discount is applied when no product is worth 100 or more.
     */
    @Test
    public void singleItemDoesNotHaveDiscountIfContainsNoProductsWorthAtLeast100() throws Exception {
        ShoppingCart shoppingCart = new ShoppingCart();
        shoppingCart.add(99);

        Assert.assertFalse(shoppingCart.hasDiscount());
    }
}
